### ISSUE TITLE
Enhance Whitespace Control

### ISSUE THREAD

**Issue by FabriceSalvaire** [pallets/jinja#712]
The aim of this issue is to discuss enhanced whitespace control.

I am using jinja to generate c++ code and I am facing problems with whitespace control (cosmetic).

The use {%- make the template difficult to read and don't solve all remaining blank lines.
And despite trim_blocks and lstrip_blocks options can improve the situation, there are cases where it is worse.

I think we need more control to handle all cases properly.

I will try to make a list of issues with trim_blocks and lstrip_blocks.

**Comment by davidism** [comment]
* I don't find `-` difficult to read.
* What blank lines are not "solved" by it?
* What cases is it worse?
* What are "all cases"? How are they "properly" handled?

This issues is too broad and vague. If you have specific issues, open issues about them with specific examples.

**Comment by ThiefMaster** [comment]
If you generate C++ (or really almost any other language) code from templates, the best solution is to not care about whitespace/indentation at all. Instead, focus on keeping the templates readable. Then, after rendering the template, run it to a reformatting tool tailored towards the code's language and let it reformat everything nicely.

**Comment by kblomqvist** [comment]
In past I also struggled a lot when using my tool called yasha to generate [this](https://github.com/kblomqvist/yasha/blob/master/tests/fixtures/nrf51.rs.expected) from [this](https://github.com/kblomqvist/yasha/blob/master/tests/fixtures/nrf51.rs.jinja). I think that Jinja's `-` does not work as expected either (can't remember in which cases).

However, today I kinda agree with ThiefMaster.

**Comment by FabriceSalvaire** [comment]
I filled this issue to open discussion on this topic, since a template engine can be used for code generation as well as for web (where cosmetic is not really a matter).

**It could be it is more a matter of documentation, tips and examples.**

I don't have problem with indentation, but remaining blank lines and over complicated templates.

**After new investigations**, it seems I have to set trim_blocks and lstrip_blocks to remove line containing jinja directives and fix remaining blank line with "-". Thus use both at same time.

Also note, I am not aware of the existence of  a compliant C++ indenter similar to C indent.

i feel there is clever strategies to write template for code generation and a tutorial on this topic could help. I found https://www.chromium.org/developers/jinja which gives many advices.
